File: app/interfaces/user_repository.py
from typing import Protocol

from models.user import User


class QueryUserParam:
    id: str
    name: str
    email: str


class QueryUserResult:
    users: list[User]
    count: int


class UserRepository(Protocol):
    def save(self, user: User):
        ...

    def findById(self, id: str) -> User | None:
        ...

    def findByEmail(self, email: str) -> User | None:
        ...

    def query(self, query: QueryUserParam) -> QueryUserResult:
        ...

File: src/main/java/com/telzz/erosync/users/UserRepository.java
package com.telzz.erosync.users;

import java.util.List;
import java.util.Optional;

import com.telzz.erosync.users.domain.User;

import jakarta.annotation.Nonnull;

public interface UserRepository {
    List<User> find();

    Optional<User> findById(String id);

    Optional<User> findByEmail(String email);

    void create(@Nonnull User user);

    void update(@Nonnull User user);
}
